File: packages/templates/src/research-agent/templates/langchain-python/tools/web_scrape.py
"""
Web Scraping Tool

Fetches and extracts text content from web pages.
"""

from langchain.tools import Tool
from typing import Dict
import requests
from bs4 import BeautifulSoup


def web_scrape(url: str, selector: str = None) -> Dict[str, str]:
    """
    Scrape content from a web page.

    Args:
        url: URL to scrape
        selector: Optional CSS selector for specific content

    Returns:
        Dictionary with url, content, title, and length
    """
    try:
        # Fetch the page
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        }
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()

        # Parse HTML
        soup = BeautifulSoup(response.text, 'html.parser')

        # Remove script and style elements
        for script in soup(["script", "style", "nav", "footer", "header"]):
            script.decompose()

        # Extract content
        if selector:
            # Use CSS selector if provided
            elements = soup.select(selector)
            content = ' '.join([el.get_text() for el in elements])
        else:
            # Extract main content
            content = soup.get_text()

        # Clean up whitespace
        lines = (line.strip() for line in content.splitlines())
        chunks = (phrase.strip() for line in lines for phrase in line.split("  "))
        content = ' '.join(chunk for chunk in chunks if chunk)

        # Limit content length
        max_length = 10000
        if len(content) > max_length:
            content = content[:max_length] + "..."

        # Extract title
        title = soup.find('title')
        title_text = title.string if title else "No title"

        return {
            "url": url,
            "title": title_text,
            "content": content,
            "length": len(content)
        }

    except requests.exceptions.RequestException as e:
        return {
            "url": url,
            "error": f"Failed to fetch {url}: {str(e)}",
            "content": "",
            "length": 0
        }
    except Exception as e:
        return {
            "url": url,
            "error": f"Failed to parse {url}: {str(e)}",
            "content": "",
            "length": 0
        }


def _web_scrape_wrapper(url: str) -> str:
    """Wrapper function for LangChain tool."""
    result = web_scrape(url)

    if "error" in result:
        return f"Error: {result['error']}"

    return f"Title: {result['title']}\n\nContent:\n{result['content']}"


# Create LangChain tool
web_scrape_tool = Tool(
    name="web_scrape",
    description=(
        "Extract text content from a web page given its URL. "
        "Useful for reading articles, blog posts, and web pages in detail. "
        "Input should be a valid URL (starting with http:// or https://). "
        "Returns the page title and main text content."
    ),
    func=_web_scrape_wrapper
)

File: packages/agents/src/tools/web-scrape.ts
import axios from "axios";
import * as cheerio from "cheerio";
import { AgentTool } from "../types";

const MAX_CONTENT_LENGTH = 15000;

/**
 * Web scraping tool that extracts clean text content from web pages.
 */
export const webScrapeTool: AgentTool = {
  name: "web_scrape",
  description:
    "Fetch and extract the text content from a web page URL. Returns the main content of the page, cleaned of HTML tags, scripts, and styles. Use this after web_search to read full articles.",
  input_schema: {
    type: "object" as const,
    properties: {
      url: {
        type: "string",
        description: "The URL to scrape",
      },
      selector: {
        type: "string",
        description:
          "Optional CSS selector to target specific content (e.g., 'article', 'main', '.content')",
      },
    },
    required: ["url"],
  },
  async execute(input: Record<string, unknown>): Promise<string> {
    const url = input.url as string;
    const selector = input.selector as string | undefined;

    try {
      const response = await axios.get(url, {
        timeout: 15000,
        maxRedirects: 5,
        headers: {
          "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
          Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        },
        responseType: "text",
      });

      const $ = cheerio.load(response.data);

      // Remove noise elements
      $("script, style, nav, footer, header, aside, iframe, noscript").remove();
      $('[role="navigation"], [role="banner"], [role="contentinfo"]').remove();
      $(".sidebar, .nav, .menu, .footer, .header, .ad, .advertisement").remove();

      // Extract content
      let content: string;

      if (selector) {
        content = $(selector).text();
      } else {
        // Try common content containers in priority order
        const selectors = ["article", "main", '[role="main"]', ".content", ".post", "#content"];
        let found = false;

        for (const sel of selectors) {
          const el = $(sel);
          if (el.length > 0 && el.text().trim().length > 100) {
            content = el.text();
            found = true;
            break;
          }
        }

        if (!found) {
          content = $("body").text();
        }
      }

      // Clean up whitespace
      content = content!
        .replace(/\s+/g, " ")
        .replace(/\n\s*\n/g, "\n\n")
        .trim();

      // Extract title
      const title = $("title").text().trim() || $("h1").first().text().trim();

      if (!content || content.length < 50) {
        return `Could not extract meaningful content from ${url}. The page may require JavaScript rendering.`;
      }

      // Truncate if too long
      if (content.length > MAX_CONTENT_LENGTH) {
        content = content.slice(0, MAX_CONTENT_LENGTH) + "\n\n[Content truncated...]";
      }

      return `**${title}**\nSource: ${url}\n\n${content}`;
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      return `Failed to scrape ${url}: ${message}`;
    }
  },
};
